src/Illuminate/Support/Traits/Macroable.php

```diff
@@ -80,10 +80,6 @@ public static function __callStatic($method, $parameters)
             ));
         }
 
-        if (static::$macros[$method] instanceof Closure) {
-            return call_user_func_array(Closure::bind(static::$macros[$method], null, static::class), $parameters);
-        }
-
```

Comment:
Does this mean `static::` can no longer be used inside static macros?

Reply:
If this change is merged then that won't be possible no.

Reply:
I have no familiarity with this macro concept, but I'm wondering if you could just rebind the scope here while leaving `$this` alone?

Reply:
@nikic hmm not sure what the method call for that would be. Afaik both `bind` and `bindTo` require `$newthis`.

Reply:
It's a bit ugly, but you could try passing `(new ReflectionFunction($closure))->getClosureThis()`.

Reply:
Ah, I was actually thinking that reflection might help here. Thanks for suggesting that 👍 

I'll try to give it a go.

Reply:
Hmm seems that didn't produce the wanted result. It still throws an error when I pass the closure. It does work when I pass in `static::class` (the class it places the macro on) but then two other tests fail.

```
public static function __callStatic($method, $parameters)
    {
        if (! static::hasMacro($method)) {
            throw new BadMethodCallException(sprintf(
                'Method %s::%s does not exist.', static::class, $method
            ));
        }

        $macro = static::$macros[$method];

        if ($macro instanceof Closure) {
            $newThis = (new ReflectionFunction(static::class))->getClosureThis();

            return call_user_func_array($macro->bindTo($newThis, static::class), $parameters);
        }

        return call_user_func_array($macro, $parameters);
    }
```

Reply:
This worked for me: https://gist.github.com/nikic/a6e97696d1351261951edeac97ef211a At least I'm not getting any test failures that look related (there are some in Carbon, but those use different code).

Reply:
@nikic I think that might be because the failing test was removed because at first we thought this couldn't be supported anymore. I've pushed a new commit with these changes and re-added the test which is now failing. (force pushed this, sorry)

https://github.com/laravel/framework/pull/29482/commits/f3b451201c54c1fe5a3a848434a92e9f444c2cb9

Failing build: https://travis-ci.org/laravel/framework/builds/571421063

Reply:
I see. The issue is that you're using `static::` rather than `self::`, and `static` is determined by `get_class($this)` if `$this` exists, with the bound scope not playing into it.

I think we need to relax this deprecation to only the cases where `$this` is actually referenced inside the closure.

Reply:
I think you should just ignore this issue for now -- I plan to relax the deprecation on the PHP side, after which (I think) no changes in Laravel should be needed anymore. You are only trying to unbind `$this` in cases where it is not used inside the closure, right?

Reply:
@nikic the test still fails if I change `static::` to `self::` [on this line](https://github.com/laravel/framework/blob/e0f075918dae4ac5975d80ade0af01dfd454883b/tests/Support/SupportMacroableTest.php#L46) it still fails. But that shouldn't be necessary since we're trying to make sure `static::` calls in closures still work.

> You are only trying to unbind $this in cases where it is not used inside the closure, right?
 
Correct yes. When using `static::` it could technically still be used in a non-static context as well but when you're calling the macro statically then there shouldn't be a concept of `$this` inside the closure.

When would these changes be implemented? We're planning to release 6.0 end of this month or early september. Depending on this I'd send the rest of this PR to the 5.8 branch already. 

Thank you for all your work on this btw :)